dotcom-rendering/src/web/components/Liveness.importable.tsx

```diff
@@ -0,0 +1,169 @@
+	pageId: string;
+	webTitle: string;
+	ajaxUrl: string;
+};
+
+const Toast = ({
+	onClick,
+	noOfNewPosts,
+}: {
+	onClick: () => void;
+	noOfNewPosts: number;
+}) => {
+	// TODO: Style and absolute position this component
+	return (
+		<nav
+			css={css`
+				position: fixed;
+				top: 20px;
+				left: 50px;
+				display: flex;
+				width: 100%;
+				align-items: center;
+			`}
+		>
+			<button onClick={onClick}>{`${noOfNewPosts} blah`}</button>
+		</nav>
+	);
+};
+
+function hydrateBlocks() {
+	console.log('hydrateBlocks');
+	// TODO Call existing hydration solution targetted to these elements
+	// Maybe we can make doHydration work for all gu-islands things
+	// where the gu-hydrated data attribute isn't set?
+}
+
+function insertNewBlocks(html: string) {
+	console.log('insertNewBlocks');
+	// TODO: This is a pure javascript action to insert the content into the dom
+	// We give it a classname causing it to be hidden by default. Unhiding it is
+	// how we achieve animation but it's dependent on scroll position
+
+	/**
+	 * TODO: We need to make the call to get new posts idempotent. As such, this
+	 * action needs to first remove any existing posts with ids the same as the
+	 * batch beforing inserting
+	 */
+
+	const latestBlock = document.querySelector('#maincontent :first-child');
+
+	if (!latestBlock) return;
+
+	latestBlock.insertAdjacentHTML('beforebegin', `<article>${html}</article>`);
+}
+
+function revealNewBlocks() {
+	console.log('revealNewBlocks');
+	// TODO: Search for and style any hidden blocks such that they become visible, fading in
+}
+
+const isServer = typeof window === 'undefined';
+
+const topOfBlog: Element | null | false =
+	!isServer && window.document.querySelector('[data-gu-marker=top-of-blog]');
+
+function topOfBlogVisible() {
+	return topOfBlog && topOfBlog.classList.contains('in-viewport');
+}
+
+if (topOfBlog) {
+	const observer = new window.IntersectionObserver(
+		([entry]) => {
+			if (entry.isIntersecting) {
+				entry.target.classList.add('in-viewport');
+				return;
+			}
+			entry.target.classList.remove('in-viewport');
+		},
+		{
+			root: null,
+			threshold: 0.1,
+		},
+	);
+
+	observer.observe(topOfBlog);
+}
+
+// TODO: Should we dismiss the toast on escape key press?
+
+export const Liveness = ({ pageId, webTitle, ajaxUrl }: Props) => {
+	const [showToast, setShowToast] = useState(false);
+	const [noOfNewPosts, setNoOfNewPosts] = useState(0);
+
+	const latestBlockId: string =
+		(!isServer &&
+			document.querySelector('#maincontent :first-child')?.id) ||
+		'';
+
+	// TODO: Read `filterKeyEvents` from the url
+	// TODO: Do we need `filterKeyEvents` and `isLivePage` here?
+	const url = `${joinUrl(
+		ajaxUrl,
+		pageId,
+	)}.json?lastUpdate=${latestBlockId}&isLivePage=true&filterKeyEvents=false`;
+
+	// useApi returns { data, loading, error } but we're not using them here
+	useApi<{
+		numNewBlocks: number;
+		html: string;
+	}>(url, {
+		refreshInterval: 10000,
+		refreshWhenHidden: true,
```

Comment:
Is this saying we're polling even when the page isn't in focus? Will we poll with the same interval when not in focus?

Reply:
Yes, `refreshWhenHidden` allows us to update the `document.title` count on the tab so readers know there's new stuff to see. I assume it polls at the same interval, that would make sense from reading the api.

Reply:
Thanks :) I was checking in the [frontend equivalent](https://github.com/guardian/frontend/blob/main/static/src/javascripts/projects/common/modules/ui/autoupdate.js#L43), I wondered if they had an increased delay for when the page is hidden? It may have been introduced to reduce load on the api, or improve some kind of device performance metric, or keep user network costs as low as possible if on mobile - it might be something that we consider too? 

PS - I hope I'm looking in the correct 'frontend equivalent', please correct me if i'm wrong :)

Reply:
There's pros and cons to a backoff strategy. Arguably, we are much more focussed on making our content 'live' now than we were back when that code was written. Backing off might reduce background noise but it also means the content stops being kept up to date which sort of breaks the contract we have with the reader (that we will keep them up to date)?

Reply:
I believe some browsers will suspend (e.g. Safari on iOS) or slow down (e.g. Chrome) JS execution in background tabs. So we may not be polling regularly (or at all) in some scenarios.

Reply:
Oh, interesting! That actually sounds ideal. I much prefer the idea of using platform configuration for this kind of thing. It will be interesting to test this but my expectation / hope is that the original polling routine is restored on focus. SWR automatically refetches on focus in any case.

Reply:
This actually tallys with what I've seen from my own testing. I've had a tab open and polling for a couple of days now and I'm not seeing wild volumes of network requests in the console but if I open it (restore focus) and watch it I see requests being made!

Of course, this is all a bit Spooky Effect From a Distance so who knows.

Reply:
I think if browsers _are_ suspending JS they should be able to re-engage it where it was left off when focus returns, otherwise they'd break all sorts of things 😬. I imagine variations of Low Power Mode also factor into this.

Having read through the full PR now, I was thinking this sort of behaviour may affect your ability to show updates in the tab title - but that's probably a small feature and it may not be impacted on all browsers?

Reply:
In comparison to Frontend, this DCR implementation keeps the tab count more up to date. If a browser does backoff the polling (I have seen this happening) then I think this is a good thing. I can also confirm that the polling resumes when you restore focus after a long time in the background.

We probably want to consider adding a field to the api response with the blog status (`isLive: boolean`) so that we can terminate polling when the blog is dead.